### PR TITLE
Create outer-loop to loop over subjects

### LINKED ISSUE #123: document plan for running pipeline on each participant/task/run
**What needs documentation and what will each item entail?**
* Plan structure of code that runs on each BIDS file from the root that the user provides in user_params
    * Diagrams, pseudocode, etc to drive development goals

as mentioned in #106

### LINKED ISSUE #138: Create outer loop to handle pipeline on subject/participant/session/run
**Describe the feature effect or output**
Outer loop to analyze intended subjects/etc based on user_params

**Which part of the pipeline does this satisfy if any**
Main pipeline

**What is the minimum for this feature to be complete?**

- user params includes info on which subject/session/run/task to analyze
- pipeline loops on info

### LINKED ISSUE #140: template functions for reref & bad-chans in main pipeline
**Describe the feature effect or output**
create template functions to populate in later pipeline steps

**Which part of the pipeline does this satisfy if any**
main pipeline

**What is the minimum for this feature to be complete?**
create empty functions for feature which has not been PR'd yet

### PR DESCRIPTION
**Pull request by F-said**
- Include section of user_params to specify subjects
- Loop over subjects, preprocessing all eeg files to close #138 
- Write intermediate eeg files to derivatives
- created template functions to close #140 
- updated README to resolve #123

### CODE REVIEW

**Comment by F-said** [comment]
closing with updated docs

File: run.py [hunk 1]
```diff
@@ -1,28 +1,34 @@
...
+# get preprocessing steps
 preprocess_steps = user_params["preprocess"]
+# get collection of subjects from user_params
+root, subjects, ch_type = load.init_data(user_params)
+
+# TODO: Should analysis be run on select tasks, runs, sessions for subjects?
+for subj in subjects:
+    subject_files = load.load_subject_files(root, subj, ch_type)
+    # get only eeg data files
+    subject_files = [s for s in subject_files if s.suffix == ch_type]
```
Comment (Jonhas) [inline run.py:18]: When we do a bids_path.match(), which is in load_subject_files, does it not already search for paths that have the same data type?

**Review by Jonhas — COMMENTED** [review]
LGTM

File: run.py (hunk 1 above)
Comment (F-said) [reply run.py:18]: It looks like the `match()` function will also load files such as 

> sub-NDARAB793GL3_ses-01_task-ContrastChangeBlock1_run-01_eeg.vhdr

I'm guessing the `datatype` parameter of `mne_bids.BIDSPath` checks for "_eeg" within the file name as opposed to the file type, so I included an additional step to remove files like these.

There's probably a better way to do this within `load_subject_files`, will look into it.

Comment (DMRoberts) [reply run.py:18]:
```suggestion
```

File: scripts/data/load.py [hunk 2]
```diff
@@ -3,7 +3,14 @@
 import json
```
Comment (DMRoberts) [inline scripts/data/load.py:3]:
```suggestion
import json
from mne_bids.config import ALLOWED_DATATYPE_EXTENSIONS
```

File: scripts/data/load.py [hunk 3]
```diff
@@ -3,7 +3,14 @@
...
 
 
-# load raw data
+def load_subject_files(path, sub, datatype):
+    bids_root = pathlib.Path(path)
+    bids_path = mne_bids.BIDSPath(subject=sub,
+                                  datatype=datatype,
+                                  root=bids_root)
+    return bids_path.match()
```
Comment (DMRoberts) [inline scripts/data/load.py:11]:
```suggestion
    matching_files = bids_path.match()
    datatype_extensions = ALLOWED_DATATYPE_EXTENSIONS[datatype]
    return [f for f in matching_files if f.extension.lower() in datatype_extensions]
```

File: scripts/data/load.py [hunk 4]
```diff
@@ -19,3 +26,8 @@ def load_param(user_param_path):
     with open(user_param_path) as fp:
         user_params = json.load(fp)
         return user_params
+
+
+def init_data(user_params):
+    init_data = user_params["load_data"]
```
Comment (DMRoberts) [inline scripts/data/load.py:32]:
```suggestion
    init_data = user_params["load_data"]
    if init_data["subjects"] == ['*']:
    init_data["subjects"] = mne_bids.get_entity_vals(init_data["root"], 'subject')
```

**Review by DMRoberts — CHANGES_REQUESTED** [review]
I added a couple suggested changes, in relation to establishing the list of data files available.

1) Getting the list of all participants in the BIDS dataset via `get_entity_vals` if requested via `*`.

2) Only retaining files that match a valid EEG datatype extension. I found that mne_bids maintains a list of valid extensions for a given datatype in the dict ALLOWED_DATATYPE_EXTENSIONS.

3) Since the extensions were filtered within load_subject_files, the line filtering on suffix could be removed from run.py.

**Comment by F-said** [comment]
Thanks for the review @DMRoberts, completed the outer loop. Now runs on:

- all participants
- a subset of participants & tasks 
- all except for a subset of participants & tasks & runs
- a subset of participants & tasks except for a subset of participants & tasks & runs

more details in #139. This feature will need a comprehensive testing module in future releases. Some test cases in #139 are probably covered but a test module would solidify consistency.

**Comment by F-said** [comment]
The informal tests that I ran this feature on are as follows:

subset of subjects and tasks 
```json
"load_data": {
      "root": "CMI/rawdata",
      "subjects": ["NDARAB793GL3"],
      "tasks": ["ContrastChangeBlock1", "RestingState"],
      "exceptions": {
        "subjects": "",
        "tasks": "", 
        "runs": ""
      },
      "channel-type": "eeg"
    }, 
```

all except for cartesian product of subjects, tasks, and runs 
```yml
"load_data": {
      "root": "CMI/rawdata",
      "subjects": ["*"],
      "tasks": ["*"],
      "exceptions": {
        "subjects": ["NDARAB793GL3"],
        "tasks": ["ContrastChangeBlock1", "RestingState"], 
        "runs": ["01"]
      },
      "channel-type": "eeg"
    }, 
```

subset of participants with exceptions
```json
"load_data": {
      "root": "CMI/rawdata",
      "subjects": ["NDARAB793GL3"],
      "tasks": ["*"],
      "exceptions": {
        "subjects": ["NDARAB793GL3"],
        "tasks": ["ContrastChangeBlock1"], 
        "runs": ["01"]
      },
      "channel-type": "eeg"
    }, 
```

all
```json
"load_data": {
    "root": "CMI/rawdata",
    "subjects": ["*"],
    "tasks": ["*"],
    "exceptions": {
      "subjects": "",
      "tasks": "", 
      "runs": ""
    },
    "channel-type": "eeg"
  }, 
```

**Review by SDOsmany — APPROVED** [review]
I like it, looks good to me as well.

**Comment by georgebuzzell** [comment]
great work, @F-said! And thank you for your help here, @DMRoberts !

Thank you for taking a look as well, @SDOsmany and @Jonhas !!

**Comment by georgebuzzell** [comment]
Move ahead with merge? @F-said

**Review by georgebuzzell — APPROVED** [review]

**Pull request merged**